pkg/database/migrate.go

```diff
@@ -195,6 +197,27 @@ func CheckAndMigrateDB(migrate bool, db *gorm.DB) error {
 			case "ac0b8b7c0014":
 				log.Info("Migrating database to FastTrackML schema 8073e7e037e5")
 				if err := db.Transaction(func(tx *gorm.DB) error {
+					// types for migration
+					type Base struct {
+						ID         uuid.UUID `gorm:"type:uuid;primaryKey"`
+						CreatedAt  time.Time
+						UpdatedAt  time.Time
+						IsArchived bool
+					}
+					type Dashboard struct {
+						Base
+						Name        string
+						Description string
+						AppID       *uuid.UUID `gorm:"type:uuid"`
+						App         App
+					}
+					type App struct {
+						Base
+						Type  string `gorm:"not null"`
+						State AppState
+					}
```

Comment:
to avoid any confusion, could you please give these types new local names that don't shadow the package types?

Reply:
also you are missing `AppState` — though you could simply give `State` its real historic type of `map[string]any`

Reply:
I agree with both of these comments, but both require interface implementation methods on the structs in order to work -- and Go doesn't like the definitions to occur here afaict.

Reply:
well maybe i can do the AppState one without the GormDataType() method. checking...

Reply:
ah right, I forgot this one was custom 🤦 — just go for `string` then!

Reply:
so I looked at this further and using those local types is super dangerous could bite us in so many places

I would recommend going for a package per migration that just contains the contents of `model.go` as it was at that time. this will be more verbose but it will ensure we have properly self-contained historical types and keep the benefits of automatic SQL generation from gorm

e.g. something like `pkg/database/migrations/e0d125c68d9a/model.go` would currently be a copy of `pkg/database/model.go`, and the migration would look like:

```go
case "5d042539be4f":
	log.Info("Migrating database to FastTrackML schema e0d125c68d9a")
	// We need to run this migration without foreign key constraints to avoid
	// the cascading delete to kick in and delete all the runs.
	if err := runWithoutForeignKeyIfNeeded(func() error {
		if err := db.Transaction(func(tx *gorm.DB) error {
			if err := tx.AutoMigrate(&e0d125c68d9a.Namespace{}); err != nil {
				return err
			}
			if err := tx.Migrator().AddColumn(&e0d125c68d9a.App{}, "NamespaceID"); err != nil {
				return err
			}
			if err := tx.Migrator().CreateConstraint(&e0d125c68d9a.Namespace{}, "Apps"); err != nil {
				return err
			}
			if err := tx.Migrator().AddColumn(&e0d125c68d9a.Experiment{}, "NamespaceID"); err != nil {
				return err
			}
			if err := tx.Migrator().CreateConstraint(&e0d125c68d9a.Namespace{}, "Experiments"); err != nil {
				return err
			}
			if err := tx.Migrator().AlterColumn(&e0d125c68d9a.Experiment{}, "Name"); err != nil {
				return err
			}
			if err := tx.Migrator().CreateIndex(&e0d125c68d9a.Experiment{}, "idx_namespace_name"); err != nil {
				return err
			}
			return tx.Model(&e0d125c68d9a.SchemaVersion{}).
				Where("1 = 1").
				Update("Version", "e0d125c68d9a").
				Error
		}); err != nil {
			return fmt.Errorf("error migrating database to FastTrackML schema e0d125c68d9a: %w", err)
		}
		return nil
	}); err != nil {
		return err
	}
```

Reply:
The proposed solution here looks really complex. Can we investigate using a technology more suited to doing migrations (ie ActiveRecord) so we aren't maintaining such baroque code?

Reply:
We should indeed look into that, but doing what I mentioned above took me 5 minutes of research and could be implemented in 30 minutes so I'd advocate to do it now so we can at least avoid the migration issues that we have repeatedly had

Reply:
What @suprjinx currently proposed will keep creating issues unfortunately (e.g. the local types reference current models and then create foreign keys on them)

Reply:
okay, I will do this -- I think we can put the migrator code in the `e0d125c68d9a` module as well, so it's more self-contained. I'll just do the two AutoMigrate cases for now

Reply:
done

Reply:
better tooling tracked in https://github.com/G-Research/fasttrackml/issues/520